expand.go

```diff
@@ -0,0 +1,217 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"io"
+
+	"golang.org/x/text/transform"
+)
+
+// expandReadCloser implements io.ReadCloser
+type expandReadCloser struct {
+	io.Reader
+	io.Closer
+}
+
+func newExpandReadCloser(rc io.ReadCloser, expandFunc func(string) (string, error)) io.ReadCloser {
+	return &expandReadCloser{
+		transform.NewReader(
+			rc,
+			&expandTransformer{expand: expandFunc}),
+		rc,
+	}
+}
+
+// expandTransformer implements transform.Transformer
+type expandTransformer struct {
+	transform.NopResetter
+	expand func(string) (string, error)
+}
+
+// First char of shell variable may be [a-zA-Z_]
+func isShellNameFirstChar(c byte) bool {
+	return c == '_' ||
+		(c >= 'a' && c <= 'z') ||
+		(c >= 'A' && c <= 'Z')
+}
+
+// Char's after the first of shell variable may be [a-zA-Z0-9_]
+func isShellNameChar(c byte) bool {
+	return c == '_' ||
+		(c >= 'a' && c <= 'z') ||
+		(c >= 'A' && c <= 'Z') ||
+		(c >= '0' && c <= '9')
+}
+
+// bytesIndexCFunc returns the index of the byte for which the
+// complement of the supplied function is true
+func bytesIndexCFunc(buf []byte, f func(b byte) bool) int {
+	for i := 0; i < len(buf); i++ {
+		if !f(buf[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Transform expands shell-like sequences like $foo and ${foo} using
+// the configured expand function.  The sequence '$$' is replaced with
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var i int
+	var j int
+
+	for i < len(src) {
+
+		if j == len(dst) {
+			return j, i, transform.ErrShortDst
+		}
+
+		end := bytes.IndexByte(src[i:], '$')
+
+		if end == -1 {
+			// src does not contain '$', copy into dst
+			cnt := copy(dst[j:], src[i:])
+			i += cnt
+			j += cnt
+			continue
+		} else if end > 0 {
+			// copy chars preceding '$' from src to dst
+			cnt := copy(dst[j:], src[i:i+end])
+			i += cnt
+			j += cnt
+
+			if j == len(dst) {
+				return j, i, transform.ErrShortDst
+			}
+
+			// fallthrough
+		}
+
+		// src[i] now points to '$', j < len(dst)
+
+		// If we're at the end of src, but we found a starting
+		// token, return ErrShortSrc, unless we're also at EOF,
+		// in which case just copy it dst.
+		if i+1 == len(src) {
+			if atEOF {
+				dst[j] = src[i]
+				i++
+				j++
+				continue
+			}
+			return j, i, transform.ErrShortSrc
+		}
+
+		// At this point we know that src[i+1] is populated.
+
+		// If this token sequence represents the special '$$'
+		// sequence, emit a '$' into dst.
+		if src[i+1] == '$' {
+			dst[j] = src[i]
+			i += 2
+			j++
+			continue
+		}
+
+		var token []byte
+
+		// Start of bracketed token ${foo}
+		if src[i+1] == '{' {
+			end = bytes.IndexByte(src[i+2:], '}')
+			if end == -1 {
```

Comment:
Once the other `end` is out of scope, you should declare this one right above
this `if` block with a slightly more descriptive name.  Something like,
`varEnd` or `tokenEnd` maybe, or a comment to go with it.

```
var tokenEnd int  // end of the $foo or ${foo}
if ... {
  // ...
  tokenEnd = ...
} else {
  // ...
  tokenEnd = ...
}
```

It makes it much clearer that `tokenEnd` is the output of the following if-
else block.

Reply:
Makes sense.